arangod/Aql/AllRowsFetcher.cpp

```diff
@@ -136,3 +152,32 @@ std::pair<ExecutionState, SharedAqlItemBlockPtr> AllRowsFetcher::fetchBlockForPa
   TRI_ASSERT(false);
   THROW_ARANGO_EXCEPTION(TRI_ERROR_NOT_IMPLEMENTED);
 }
+
+std::pair<ExecutionState, ShadowAqlItemRow> AllRowsFetcher::fetchShadowRow(size_t atMost) {
+  // We are required to fetch data rows first!
+  TRI_ASSERT(_aqlItemMatrix != nullptr);
+
+  ExecutionState state = _upstreamState;
+  ShadowAqlItemRow row = ShadowAqlItemRow{CreateInvalidShadowRowHint{}};
+  if (!_aqlItemMatrix->stoppedOnShadowRow() || _aqlItemMatrix->size() == 0) {
+    // We ended on a ShadowRow, we are required to fetch data.
+    state = fetchData();
+    if (state == ExecutionState::WAITING) {
+      return {state, row};
```

Comment:
It might be more transparent (but quite verbose) to
```
return {ExecutionState::WAITING, ShadowAqlItemRow{CreateInvalidShadowRowHint{}};
```

(I have not really made up my mind about this one yet)